astropy/utils/shapes.py

```diff
@@ -333,11 +352,17 @@ def __array_function__(self, function, types, args, kwargs):
 
 
 class IncompatibleShapeError(ValueError):
-    def __init__(self, shape_a, shape_a_idx, shape_b, shape_b_idx):
+    def __init__(
+        self,
+        shape_a: tuple[int, ...],
+        shape_a_idx: int,
+        shape_b: tuple[int, ...],
+        shape_b_idx: int,
+    ) -> None:
         super().__init__(shape_a, shape_a_idx, shape_b, shape_b_idx)
 
 
-def check_broadcast(*shapes):
+def check_broadcast(*shapes: tuple[int, ...]) -> tuple[int, ...]:
```

Comment:
it might be useful to have a `Shape` type alias defined here or in a `typing` module

Reply:
`tuple[int, ...]` is a very simple annotation and I don't think we'd gain much in readability by introducing an alias for it. However, if `numpy.typing` already has a type for shapes then it might be a good idea to use that. I am not aware of such a type, but I'm not as familiar with `numpy.typing` as I'd like to be.

Reply:
See https://github.com/numpy/numpy/blob/2d2ff2501ad8a558c9ea402bf7b1a48e3f676fed/numpy/_typing/_shape.py#L4

May be worth checking with numpy to what extent these definitions are meant to be private...

Reply:
The shorter and simpler a type is the less we gain by having an alias for it. `tuple[int, ...]` is both short and simple, so using the private `numpy.typing._Shape` doesn't seem to be worth it. `SupportsIndex | Sequence[SupportsIndex]` is still simple, but it is more verbose, so using an alias is more appealing, but `_ShapeLike` is private too. In practice `int | tuple[int, ...]` (or possibly `int | Sequence[int]`) should be a good enough substitute for the actual shape-like type in most cases.

Reply:
Here, I feel we should definitely not let perfect be the enemy of good enough. `tuple[int, ...]` is very clear (and plain `int` is not allowed here). I guess in the end it matters if type checking software is going to have a problem with someone calling `check_broadcast((np.int64(1),), (10,))` or `check_broadcast([1], [10])` (both of which work fine). So, in that sense `Sequence[SupportsIndex]` would be more strictly correct (but arguably less clear for the human reader...)